### PR TITLE
[WOO POS] update UI for payment processing content in checkout

### PR DESCRIPTION
**Pull request by bozidarsevo**
#13120

## Description

- Updated the components used to show payment processing statuses
- Status icons are missing for now, that will be done in another PR, for now I put a plan rectangle in place where we will show a status icon (as per design)
- When we start processing the payment the whole checkout screen expands to full width

In next PRs:
- animated icons
- animated background gradient
- more error handling and unhappy path handling

## Testing information

- Enter POS
- Add few items in the cart
- Checkout
- Do the whole payment process and check that the payment processing content is shown properly

## Screenshots

updatedpayment

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift]: 💭 for the 3 `POSCardPresentPaymentMessageView` use cases in this file, how about creating a separate view for each case so that they can be customized more easily? For example, if we want to add some animation only in the success case. If we really want to reuse some SwiftUI components like the title, they can still be shared within their own views.

I also felt like it's a bit odd to reuse `POSCardPresentPaymentMessageView` with other use cases in the PR to show buttons and progress view. Having separate views for different purposes and reusing only what's common is what I tend to follow using SwiftUI.

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift]: nit: can move the dimensions to Layout constants

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift]: nit: for the font, is it possible to map to the semantic style like `body`?

**Review by jaclync — COMMENTED** [review]
Shared some thoughts about having separate views for different use cases instead of having one view with various optional parameters 💭 

While we're on UI updates for the payment states, potential updates for the UI to match design (ready for dev M1) TfaZ4LUkEwEGrxfnEFzvJj/WooCommerce-POS?node-id=1439-73823&t=uZRnmnPn92HHO9W5-0:

design | PR branch
-- | --
Screenshot 2024-06-27 at 2 39 29 PM | 

Maybe we can also:
- Update the title "Reader is ready" to "Ready for payment"
- Remove the cancel CTA 

Which is a subtask of https://github.com/woocommerce/woocommerce-ios/issues/13071.

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift
Comment (bozidarsevo) [reply WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift]: There were/are special views for a lot of cases but since they show the content in the same layout/design I was thinking of having one view to show that data so we do not have 10 places to update if we change a color or a font.
I was even thinking of removing `PointOfSaleCardPresentPaymentTapSwipeInsertCardMessageView` and similar views and just use `POSCardPresentPaymentMessageView`.

There is no rush with this PR so I will maybe wait when I get those animated icons and see what would be the best approach and how many view structs to have.

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift
Comment (bozidarsevo) [reply WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift]: I could, but I do not know what are "rule" is, to either follow strictly Figma font sizes or use the semantic ones?

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift]: I agree that there are UI elements that can be reused like the title and message for states that share the same layout. For `POSCardPresentPaymentMessageView`, could we try moving the optional image, circle, progress view, and buttons out to separate views, and replace the image/circle/progress part with a generic `topContent: View` and button container with a generic `bottomContent: View`? This way, if we touch the layout of the optional image/circle/progress/buttons for a subset of states, we don't have to worry about affecting the UI of other states.

Maybe we can have a third/fourth opinion from @iamgabrielma and @staskus as well! I think this affects how we develop UI down the road and is worth discussing as a team.

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift]: If it's not specified in the design, I used to map it manually with the typography guide pzlDWpqXNhwq8kzOGMmMEh/Woo-Mobile---iOS-Component-Library?node-id=2249-3664&t=gbYHZGnwJwDEaM3S-0 (if the link is hard to work with, it's also linked in P91TBi-3PM-p2 > Component Libraries). But with the custom font that's used in the design, the mapping might not work. Just a non-blocking nit suggestion.

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift
Comment (staskus) [reply WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift]: > There is no rush with this PR so I will maybe wait when I get those animated icons and see what would be the best approach and how many view structs to have.

I agree with the iterative process, right now for me it's difficult to say what an ideal setup for these views would look like, depending on how many similarities and differences we'll have. Choosing to actively refactor while developing could help us find the balance between customization and reusability. 

> I was thinking of having one view to show that data so we do not have 10 places to update if we change a color or a font.

Yes, agree that having a completely different view for every different configuration won't make sense if we want to share some if not most attributes between different views.

> ..could we try moving the optional image, circle, progress view, and buttons out to separate views <..> This way, if we touch the layout of the optional image/circle/progress/buttons for a subset of states, we don't have to worry about affecting the UI of other states.

Something like that seems reasonable to me. The main view could share certain styling attributes with child views so we wouldn't need to have 10 places to update if we change a color or a font. Having 4 optional members and one non-optional generates lots of possible permutations of this view. I imagine the number of members will expand as views become more and more complex. We would notice the complexity if we tried to create previews that would cover all the possible scenarios that this UI should theoretically handle.

**Comment by bozidarsevo** [comment]
Tnx @jaclync and @staskus for feedback! I will wait a day or two for the final changes, just to wait for the model changes to be merged so I am sure my merge will not break anything!

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/POSCardPresentPaymentMessageView.swift:3]: I think we need to iterate on this model (totally fine on a different PR): Going by its properties, we're allowing the view model to be initialized with all properties being nil or empty, however, all its properties are constants so we cannot re-assign them or perform property injection after it has been initialized. 

I think this could indicate that we're trying to reuse it more than we should and some of its properties should be enforced on initialization by removing the Optional and default values. Perhaps even managing all the properties privately/internally only and just exposing an enum with the type of viewmodel we should initialize 🤔

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/PointOfSaleCardPresentPaymentTapSwipeInsertCardMessageView.swift
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/Reader Messages/PointOfSaleCardPresentPaymentTapSwipeInsertCardMessageView.swift]: ❓ Should we move the `pos_ready_for_payment` string to some sort of extension like we do for `UIImage+Woo`?

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift
Comment (iamgabrielma) [reply WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift]: > There is no rush with this PR so I will maybe wait when I get those animated icons and see what would be the best approach and how many view structs to have.

💯 

> Having 4 optional members and one non-optional generates lots of possible permutations of this view. I imagine the number of members will expand as views become more and more complex

Agree with this, I've also dropped a similar [comment](https://github.com/woocommerce/woocommerce-ios/pull/13166#pullrequestreview-2152591848) for the viewmodel where we might be trying to handle too many cases for the sake of reusability. Jaclyn's suggestion of having some main view with only the reusable components, and move the optional ones to sub-views feels a good approach to explore, this could even be managed internally by the view model we initialize the subviews with, so it wouldn't need to be exposed.

**Comment by iamgabrielma** [comment]
>  I will wait a day or two for the final changes, just to wait for the model changes to be merged so I am sure my merge will not break anything

Looks good! Thanks for holding this PR for a bit, the [remaining changes](https://github.com/woocommerce/woocommerce-ios/pull/13211) should ideally be merged today and unblock the rest of the work 🙇

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift
Comment (bozidarsevo) [inline WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift]: I am having a hard time naming this property.
Basically this boolean is used to determine if the whole totals/payments view is full screen (cart is not showed).
Anybody has any idea for a better name or is this ok?

Comment (jaclync) [reply WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift]: How about something like `isTotalsViewFullScreen`? A comment about the variable like what you described would be helpful too.

File: WooCommerce/Resources/en.lproj/Localizable.strings
Comment (jaclync) [inline WooCommerce/Resources/en.lproj/Localizable.strings:7523]: nit: how about using a different key to replace the "preparingCardReader" part for this string, since the message doesn't seem to be about "preparing card reader" anymore? We wouldn't need to update the strings file anymore.

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/Reader Messages/PointOfSaleCardPresentPaymentSuccessMessageViewModel.swift
Comment (jaclync) [inline WooCommerce/Classes/POS/Presentation/Card Present Payments/Reader Messages/PointOfSaleCardPresentPaymentSuccessMessageViewModel.swift]: nit: could we move the image definition to `UIImage+Woo.swift` so that we can also ensure it's non-nil in the unit tests in `IconTests`?

Also, it would be nice to follow the naming convention in the current images in Images.xcassets using dashes instead of underscores.
```suggestion
    // Moves this to UIImage+Woo to return `UIImage(named: "pos-payment-successful")` and uses the UIImage reference here
    let imageName = "pos-payment-successful"
```

File: WooCommerce/Resources/en.lproj/Localizable.strings
Comment (bozidarsevo) [reply WooCommerce/Resources/en.lproj/Localizable.strings:7523]: I left it like that since it is in the "preparing card reader" stage, where the title is "Preparing card reader" and this is the message below the title.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/212">19.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/Reader Messages/PointOfSaleCardPresentPaymentSuccessMessageViewModel.swift
Comment (bozidarsevo) [reply WooCommerce/Classes/POS/Presentation/Card Present Payments/Reader Messages/PointOfSaleCardPresentPaymentSuccessMessageViewModel.swift]: I was thinking of just having the name so I can use it like `Image(imageName)` and avoid having UIKit code in the `SwiftUI` view. Using the `UIImage+Woo` property would help to avoid typos with the strings.
Maybe I just have it like this:
```
static var posPaymentSuccessfulImageName: String {
 return "pos-payment-successful"
}

static var posPaymentSuccessfulImage: UIImage {
 UIImage(named: posPaymentSuccessfulImageName)!
}
```

And that way the string is used just in one place and we can reference the name or the `UIImage as we need it.

Comment (bozidarsevo) [reply WooCommerce/Classes/POS/Presentation/Card Present Payments/Reader Messages/PointOfSaleCardPresentPaymentSuccessMessageViewModel.swift]: Hm, even with that I would have it `let imageName = UIImage.posPaymentSuccessfulImageName` and would basically have to import UIKit in the view model files.

I think I will add image names to a String extension and have it like that to avoid extra UIKit imports

File: WooCommerce/Classes/POS/Presentation/Card Present Payments/Reader Messages/PointOfSaleCardPresentPaymentPreparingForPaymentMessageViewModel.swift
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/Card Present Payments/Reader Messages/PointOfSaleCardPresentPaymentPreparingForPaymentMessageViewModel.swift:17]: np; 
A comment may need to be updated as well.

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift
Comment (staskus) [inline WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift:23]: https://github.com/woocommerce/woocommerce-ios/assets/4062343/c4a9568d-70ec-43d3-b941-88e2c8f04439

It may not be related to these changes but I noticed that the success view is briefly visible when starting another payment. Likely `cardPresentPaymentInlineMessage` is not reset. If `TotalsViewModel` _holds_ a state for a specific payment, we could consider creating a new one every time a new payment starts and avoid such a category of issues. Let me know if I should create an issue for it.

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift
Comment (staskus) [inline WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:153]: Observation:

There are errors that appear on full screen and there are errors that appear in half screen depending on the state when they appear (processing payment vs preparing reader).

Given that, we should either show errors in a consistent view-form or be mindful of that and make sure that our state/error views work both in full screen and half screen

image

**Review by staskus — APPROVED** [review]
Works well. The changes in the code are understandable and the scenarios that I managed to test work as expected:

✅ Reader Disconnected
✅ Reader Connected
✅ cancelledOnReader (couldn't test since I don't have a reader but checked on SwiftUI preview)
✅ Progress view while preparing payment
✅ Icons for some of the statuses
✅ Retryable errors
✅ Non-retryable errors
✅ isTotalsViewFullScreen (full screen for processing and paid states)

I left only a couple of observations for the future.

File: WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift
Comment (bozidarsevo) [reply WooCommerce/Classes/POS/Presentation/CardReaderConnection/UI States/PointOfSaleCardPresentPaymentInLineMessage.swift:23]: Yeah, I noticed that few days ago. Created https://github.com/woocommerce/woocommerce-ios/issues/13275

File: WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift
Comment (bozidarsevo) [reply WooCommerce/Classes/POS/ViewModels/PointOfSaleDashboardViewModel.swift:153]: Tnx! Better error handling will be done in next PRs!

**Pull request merged**
